### DISCUSSION TITLE
Materials in FV volume kernel

### DISCUSSION THREAD

**Discussion by heinono1** [idaholab/moose#22350] [Q&A General] [answered]
I am setting up some kernels for a FV problem. In one kernel, I need to access materials properties for a volume contribution to the residual. So I have InputParameters:
```
InputParameters
MyFVVolumeKernel::validParams()
{
InputParameters params = FVElementalKernel::validParams();
..
params.addRequired<MooseFunctorName>("Km", "materials property");
....
return params;
}

and later

MyFVVolumeKernel::MyFVVolumeKernel(const inputParameters & parameters):FVElementalKernel(parameters),
...
_Km(getFunctor<ADRreal>("Km"),
..

and for the residual
ADReal
MyFVVolumeKernel::computeQPResidual()
{
return _Km[_qp]*u[_qp];
}
```
The problem I have seems to be that I cannot evaluate _Km at quadrature points - the error when I compile is
"no match for 'operator[]'"

How can I fix this?
Thanks!

**Comment by GiudGiud** [comment, marked as answer]
Hello

Functions and functors cant be access with []. Functions use .value(), functors use ()
We access functors by passing arguments, see some examples in Navier Stokes or in the framework.
A Qp, a face, an element or a position are all accepted arguments

https://github.com/idaholab/moose/blob/next/framework/src/auxkernels/FunctorElementalAux.C

Guillaume

**Comment by heinono1** [comment]
Ah, got you.

Thanks, Guillaume!

On Mon, Oct 10, 2022 at 3:07 PM Guillaume Giudicelli <
***@***.***> wrote:

> Hello
>
> Functions cant be access with [].
> We access functors by passing arguments, see some examples in Navier
> Stokes or in the framework.
> A Qp, a face, an element or a position are all accepted arguments
>
>
> https://github.com/idaholab/moose/blob/next/framework/src/auxkernels/FunctorElementalAux.C
>
> Guillaume
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/22350#discussioncomment-3842472>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEF4KAT2B4PCZN6JVZNDWCRZRZANCNFSM6AAAAAARBT3F2Q>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***
